### ISSUE TITLE
gcovr 51.: AssertionError: assert self.lineno == other.lineno

### ISSUE THREAD

**Issue by simonspa** [gcovr/gcovr#586]
**Describe the bug**
Running the latest gcovr 5.1 on our otherwise unchanged code base aborts with:

```
(ERROR) Uncaught EXCEPTION
Traceback (most recent call last):
  File "~/.local/bin/gcovr", line 8, in <module>
    sys.exit(main())
  File "~/.local/lib/python3.9/site-packages/gcovr/__main__.py", line 308, in main
    collect_coverage_from_gcov(covdata, options)
  File "~/.local/lib/python3.9/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "~/.local/lib/python3.9/site-packages/gcovr/workers.py", line 170, in __exit__
    self.wait()
  File "~/.local/lib/python3.9/site-packages/gcovr/workers.py", line 161, in wait
    raise self.exceptions[0][1]
  File "~/.local/lib/python3.9/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "~/.local/lib/python3.9/site-packages/gcovr/workers.py", line 161, in wait
    raise self.exceptions[0][1]
  File "~/.local/lib/python3.9/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "~/.local/lib/python3.9/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "~/.local/lib/python3.9/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "~/.local/lib/python3.9/site-packages/gcovr/gcov.py", line 152, in process_gcov_data
    covdata.setdefault(key, FileCoverage(key)).update(coverage)
  File "~/.local/lib/python3.9/site-packages/gcovr/coverage.py", line 374, in update
    self.function(fct_name).update(fct_cov)
  File "~/.local/lib/python3.9/site-packages/gcovr/coverage.py", line 200, in update
    assert self.lineno == other.lineno
AssertionError
```

Running `gcovr` with its verbosity flag, gives some hint on which file it aborts:

```
(DEBUG) Processing file: ~/software/allpix-squared/build/src/modules/GDMLOutputWriter/CMakeFiles/AllpixModuleGDMLOutputWriter.dir/__/__/core/module/dynamic_module_impl.cpp.gcda
(DEBUG) Running gcov: 'gcov ~/software/allpix-squared/build/src/modules/GDMLOutputWriter/CMakeFiles/AllpixModuleGDMLOutputWriter.dir/__/__/core/module/dynamic_module_impl.cpp.gcda --branch-counts --branch-probabilities --demangled-names --hash-filenames --object-directory ~/software/allpix-squared/build/src/modules/GDMLOutputWriter/CMakeFiles/AllpixModuleGDMLOutputWriter.dir/__/__/core/module' in '~/software/allpix-squared/build'
(DEBUG) Finding source file corresponding to a gcov data file
  currdir      ~/software/allpix-squared/build
  gcov_fname   ~/software/allpix-squared/build/dynamic_module_impl.cpp##6fb4120403a0b1da6447f55ee298eebf.gcov
  source_fname ~/software/allpix-squared/build/src/modules/GDMLOutputWriter/CMakeFiles/AllpixModuleGDMLOutputWriter.dir/__/__/core/module/dynamic_module_impl.cpp.gcda
  root         ~/software/allpix-squared
  fname        ~/allpix-squared/src/core/module/dynamic_module_impl.cpp
(DEBUG) Parsing coverage data for file ~/software/allpix-squared/src/core/module/dynamic_module_impl.cpp
Traceback (most recent call last):
```

I do not know which files you need to further trace the issue:

* The original code is here: https://github.com/allpix-squared/allpix-squared/blob/master/src/core/module/dynamic_module_impl.cpp
* The `.gcov` file is attached here: https://pastebin.com/ZsyJ39rd

**Expected behavior**
`gcovr` should parse the files without issue, as was the case with previous versions.

**Comment by latk** [comment]
Thank you for this excellent report.

The present source file has a curios pattern that has different definitions for a function depending on preprocessor `#define`s. Basically:

```c++
#if ALLPIX_MODULE_UNIQUE
    Module* allpix_module_generator(Configuration& config, Messenger* messenger, GeometryManager* geo_manager) {
        ...
    }
#else
    Module*  allpix_module_generator(Configuration& config, Messenger* messenger, std::shared_ptr<Detector> detector) { // NOLINT
        ...
    }
#endif
```

You also seem to be merging coverage data from compiling this file with different preprocessor options. Thus, we have two different functions with the same name:

* compilation unit 1: function `allpix_module_generator()` on line 54
* compilation unit 2: function `allpix_module_generator()` on line 76

Gcovr 5.1 adds support for function coverage. Usually, execution counts for functions can be merged across different compilation units, since they refer to the same function. But here, we have two **different functions** with the **same name**, and it would be inappropriate to just merge coverage data for them – just because they have the same name does not mean that these functions are in any way related.

The solutions I see here:

* Either gcovr uses both function name + line number to disambiguate functions so that it supports multiple identically named functions in the same data set.
* Or, gcovr merges function coverage data based solely on the name (simply by skipping that assert).
* Or, this use case remains unsupported.

I'm tending towards the third solution since it preserves the status quo, and since it is the only one that will alert a user to potential corruption. It is generally possible to write the code with only a single function definition. The problem will also disappear if the file is only compiled once.

I know I'm biased because I want to keep gcovr simple, but in my experience it's often best to avoid lots of ifdefs, especially since this also prevents some degree of type-checking during compilation. I'd rather write that kind of code as:

```c++
    Module* allpix_module_generator_unique(Configuration& config, Messenger* messenger, GeometryManager* geo_manager) {
        ...
    }

    Module*  allpix_module_generator_shared(Configuration& config, Messenger* messenger, std::shared_ptr<Detector> detector) { // NOLINT
        ...
    }

#if ALLPIX_MODULE_UNIQUE
   using allpix_module_generator = allpix_module_generator_unique;
#else
   using allpix_module_generator = allpix_module_generator_shared;
#endif
```

buuut I've noticed that this function relates to dynamic loading so you probably need to ensure a particular symbol name.

Thoughts on this? How would you think function coverage should be aggregated?

**Comment by simonspa** [comment]
Hi @latk 

thank you for the blazingly fast answer and your detailed explanation!

The weird code abusing the same function name for different footprints is indeed related to dynamic loading. The file in question is included in different compilation targets, serving as the boilerplate code for dynamic libraries. These libraries are loaded and objects are created through the same symbol name, but with different parameters, depending on the underlying class. Odd in code but very flexible otherwise. ;)

Hence, my biased-towards-our-dynamic-loading preference would of course be the distinction between the functions based on both name and line, but also simply skipping the assertion would be fine with me.

**Comment by latk** [comment]
Ok, I'll think about a way to skip this check on an opt-in basis, for example with a new `--allow-duplicate-functions` flag. I'd really like to keep an assertion because it does point to a problem in many cases, but your code pattern is somewhat common and deserves appropriate support.

**Comment by Spacetown** [comment]
I've checked which line is used if the definition is written across several lines and clang-13 uses the line of the function name.
My preferred solution is to fix this on project level instead of doing a workaround in gcovr. Can you try to define the function in this way?

```C
    Module* allpix_module_generator(
#if ALLPIX_MODULE_UNIQUE
        Configuration& config, Messenger* messenger, GeometryManager* geo_manager
#else
        Configuration& config, Messenger* messenger, std::shared_ptr<Detector> detector
#endif
       ) {
        ...
    }
```
or
```C
    Module* allpix_module_generator(
        Configuration& config, Messenger* messenger, 
#if ALLPIX_MODULE_UNIQUE
        GeometryManager* geo_manager
#else
        std::shared_ptr<Detector> detector
#endif
       ) {
        ...
    }
```

**Comment by latk** [comment]
@pustekuchen91 Thank you for that example! Which compiler are you using?

Constexpr-lambdas are certainly an interesting pattern that I have not seen before, but I'm not quite sure why that would break things in the code that you've shown. But I think that with similar code, Clang could perhaps be tricked into assigning the same (mangled) function name to different lambdas in different compilation units.

In any case, this is a piece of evidence suggesting that differing line numbers for the “same” function occur naturally in some projects and that I'll have to think about how that can be handled in gcovr. Today I have merged some code that makes it *possible* to address this issue, but no concrete fix exists yet.

**Comment by pustekuchen91** [comment]
Hi @latk, 

i'm using g++ 10, running in a [conan docker container](https://hub.docker.com/r/conanio/gcc10/).

**Comment by simonspa** [comment]
> Can you try to define the function in this way?

While in principle this might work, the result is not very readable in the source code. If you check the linked source file, this would mean a clutter-`ifdef`-mix of documentation, declaration, definition to work around this... If possible I'd prefer to keep the source code clean an readable for humans and tell the tools to work around what they regard as an issue :slightly_smiling_face:

**Comment by Spacetown** [comment]
Is reusing a name with a different prototype clean? What about using a typedef for the third argument? Than there is only one place with the encapsulation.
For us that's an issue because it's really a different function.

**Comment by Pesa** [comment]
> Gcovr 5.1 adds support for function coverage. Usually, execution counts for functions can be merged across different compilation units, since they refer to the same function. But here, we have two **different functions** with the **same name**, and it would be inappropriate to just merge coverage data for them – just because they have the same name does not mean that these functions are in any way related.

Just using the function name as identifier will break overloaded functions, which are very common in C++ code...

> 
> Thoughts on this? How would you think function coverage should be aggregated?

How about using the mangled symbol name to identify a function?

**Comment by latk** [comment]
> How about using the mangled symbol name to identify a function?

@Pesa We are not just using the function “name”, which, as you correctly point out, is meaningless in a C++ context.

Gcovr requests demangled names if possible, but these will also be unique. For all practical purposes, the demangled names are just as good as the mangled symbols. I have yet to see an example where GCC's demangled names caused problems. In particular, they also contain all the type information that is necessary for overload resolution. For example, see this [example on the Godbolt compiler explorer](https://godbolt.org/z/4PK1WW1Yc) showing demangled names including templates and CV-qualified pointers. E.g. a function declared as `void foo(const int*)` will have mangled symbol `_Z3fooPKi` which demangles as `foo(int const*)`.

The only caveat I've learned about so far is the names assigned by Clang to lambdas, which are unique within a compilation unit but not stable between multiple compilations of the same source (see above discussion).

It is also worth mentioning that there is no relationship between function coverage and other metrics. In particular, gcovr aggregates line and branch coverage across all template specializations of a function, but counts each specialization as a separate function for function coverage purposes.

**Comment by latk** [comment]
In other news, I've been experimenting with supporting different function coverage merging modes, but this is rather tricky since this implies dealing with a set of objects that change their concept of equality dynamically. I'll have to think about this further. It is not impossible, just more tricky than expected.

**Comment by Pesa** [comment]
Thanks for the clarification. I think I got confused by this earlier comment:
> Thus, we have two different functions with the same name:
> compilation unit 1: function allpix_module_generator() on line 54
> compilation unit 2: function allpix_module_generator() on line 76

Using the _full_ demangled function name (i.e. including parameters) should indeed be fine, and should be essentially equivalent to the mangled symbol name.

**Comment by nmoinvaz** [comment]
I am able to reproduce this assert in [googletest](https://github.com/google/googletest) in gtest.h with the `RUN_ALL_TESTS` function.

<img width="487" alt="image" src="https://user-images.githubusercontent.com/1364432/167539141-4202984d-f702-45e1-8ff3-3a1ddd6dae0c.png">

<img width="387" alt="image" src="https://user-images.githubusercontent.com/1364432/167539481-4daa6bf9-4db2-44ea-85e2-3ed50b34dde9.png">

The error only happens when using `--gcov-executable "llvm-cov-11 gcov"`. When using `--gcov-executable "llvm-cov-10 gcov"` the problem does not exist.

**Comment by Spacetown** [comment]
@latk Was this intended to turn of the assert by default in the new parser?
https://github.com/gcovr/gcovr/blob/933731ad0ad8401f750780a7f0ec95033a60cf67/gcovr/gcov_parser.py#L640-L644

**Comment by latk** [comment]
@Spacetown Honestly, I don't quite remember. I know I added the `MergeOptions()` concept to make it *possible* to merge otherwise conflicting data. But so far, that code is *almost* bug-compatible with the previous implementation:

https://github.com/gcovr/gcovr/blob/836839bd1350442b0908a7908927d9997f65a8a7/gcovr/gcov_parser.py#L616-L618

(There is a minor difference in that the old parser assigns linenos in an unspecified order, whereas the new `merge_function()` in `ignore_function_lineno` mode uses the maximum lineno for that function.)

However, the `ignore_function_lineno` mode is not enabled at two important points:

* when merging coverage data from two gcov files as part of normal execution, or when using multiple threads
* when merging coverage data from JSON files

Thus, the assertion is still triggered.

The current `ignore_function_lineno` mode also doesn't address the use case from this issue: Functions of the same name are necessarily merged into one coverage statistic, whereas this issue would benefit from being able to treat functions with the same name but on different lines as distinct functions.

Ideally, there would be another merging mode to treat functions with different linenos as distinct entities. And these merging modes would be exposed on the command line. But that is where I hit a wall and didn't make further progress.

If you want to continue work on this, it could make sense to expose function merging modes as a command line option (not as a yes/no flag, but as a named option making it possible to implement more modes). Then users could at least disable the assertion.

**Comment by Spacetown** [comment]
No I'm really confused. 😞 
The function following function is called in two place.
https://github.com/gcovr/gcovr/blob/933731ad0ad8401f750780a7f0ec95033a60cf67/gcovr/merging.py#L185

- When merging JSON files the default merge option is used (`ignore_function_lineno = False`) which triggers the assertion.
  https://github.com/gcovr/gcovr/blob/933731ad0ad8401f750780a7f0ec95033a60cf67/gcovr/writer/json.py#L191
- When parsing GCOV files (`ignore_function_lineno = True`) which doesn't trigger the assertion and takes the max line number.
  https://github.com/gcovr/gcovr/blob/933731ad0ad8401f750780a7f0ec95033a60cf67/gcovr/gcov_parser.py#L640

For me your description doesn't fit to the current state.

> If you want to continue work on this, it could make sense to expose function merging modes as a command line option (not as a yes/no flag, but as a named option making it possible to implement more modes). Then users could at least disable the assertion.

This make sense with a default of `None`. If not specified, the default depends on the usecase.

**Comment by gparmeggiani** [comment]
Hi, 
I'm also facing this issue.

Is there a workaround (maybe via configuration) that I can use to ignore this function line number check?

**Comment by Spacetown** [comment]
> Hi, I'm also facing this issue.
> 
> Is there a workaround (maybe via configuration) that I can use to ignore this function line number check?

Are you also using the preprocessor in the function definition?

**Comment by gparmeggiani** [comment]
yes, exactly.
For now I patched the `merging.py` file so that the default value of `ignore_function_lineno` is `True`.  
Let me know if there is a better workaround.

Happy to help with more details, in case you need them to fix the issue

**Comment by yefen123** [comment]
Hi, I'm also facing this issue and I have created a "question" : **https://github.com/gcovr/gcovr/issues/673** 3 days ago.
I set "ignore_function_lineno == True" in /proj/crbs/zyexfen/virtual_env/lib/python3.7/site-packages/gcovr/merging·py , 
![563e81edd2c2a5623646a0751763849](https://user-images.githubusercontent.com/111563279/191639625-81ab4fb3-e46a-4763-a3c4-2b7f0415f365.png)
**but report another error:**
Traceback (most recent call last):

File "/proj/crbs/zyexfen/yfenv/bin/gcovr", line 7, in <module> 
  from gcovr. main import main
File "/proj/crbs/zyexfen/yfenv/lib/python3.7/site-packages/gcovr/_main_.py", line 38, in <module> 
  from.gcov import (
File "/proj/crbs/zyexfen/yfenv/lib/python3.7/site-packages/gcovr/gcov.py", line 31, in <module> 
  from.gcov_parser import parse metadata, parse_coverage,ParserFlags
File"/proj/crbs/zyexfen/yfenv/lib/python3.7/site-packages/gcovr/gcov_parser.py", line 57, in <module> 
  from.merging import (
File "/proj/crbs/zyexfen/yfenv/lib/python3.7/site-packages/gcovr/merging.py",line 272 
  left.count +=right.count
IndentationError: expected an indented block 

So can you tell me how to fix the issue? Thanks in advance.

**Comment by Spacetown** [comment]
> So I wonder when can the formal solution be released?

The formal solution is to define a function with the same name always at the same line.

For now we have to finish the open PRs first. After that we can think about a merge option.

**Comment by mawillers** [comment]
FWIW, I've also just run into this issue with gcovr 5.2. My use case is basically this C++14 code:
```
#include <type_traits>

template <typename T>
class S
{
#if defined(USE_SAFE_FALLBACK)
        static constexpr bool isSafe = true;
#else
        static constexpr bool isSafe = false;
#endif

        template <typename U = T, std::enable_if_t<std::is_same<U, char>::value && isSafe, int> = 0>
        static int impl(U) { return 1; }

        template <typename U = T, std::enable_if_t<!(std::is_same<U, char>::value && isSafe), int> = 0>
        static int impl(U) { return 1; }

public:
        static int f(T t) {
                return impl(t);
        }
};
```
i.e. some SFINAE of type traits combined with a preprocessor symbol. I can probably express this differently, but it's certainly code that is not unusual nowadays.

I then want to unit test & run coverage for both the "isSafe" settings, so I'm building two test executables from the same source, once with and once without USE_SAFE_FALLBACK.

The traceback looks pretty much like the above one. If you're interested, I'm happy to provide it, as well as the complete code example.
Setting that `ignore_function_lineno` flag to `True` appears to fix it for me, too.

**Comment by Spacetown** [comment]
@mawillers Following definition should fix your problem:
```c
        static constexpr bool isSafe =
#if defined(USE_SAFE_FALLBACK)
            true;
#else
            false;
#endif

**Comment by mawillers** [comment]
@Spacetown, I've tried that, but it shows the same crash.

What _does_ work is, for instance, changing the template argument SFINAE into e.g. a return type SFINAE:
```
        template <typename U = T>
        static
        std::enable_if_t<std::is_same<U, char>::value && isSafe, int>
        impl(U) { return 1; }

        template <typename U = T>
        std::enable_if_t<!(std::is_same<U, char>::value && isSafe), int> 
        static
        impl(U) { return 1; }
```

**Comment by Spacetown** [comment]
Ok, that was a misunderstanding from my side.
Can you check the function names generated in the gcov files (option `--keep`)? In the case of the error they should be the same but how can this be accepted by the compiler?

**Comment by mawillers** [comment]
The function _names_ are the same, but their signature is different, as one of the functions is for T==char, whereas the other one is for T!=char.
This would be the interesting part, I guess:
```
        -:    8:template <typename T>
        -:    9:class S
        -:   10:{
        -:   11:#if defined(USE_SAFE_FALLBACK)
        -:   12:        static constexpr bool isSafe = true;
        -:   13:#else
        -:   14:        static constexpr bool isSafe = false;
        -:   15:#endif
        -:   16:
        -:   17:        template <typename U = T, std::enable_if_t<std::is_same<U, char>::value && isSafe, int> = 0>
function int S<char>::impl<char, 0>(char) called 1 returned 100% blocks executed 100%
        1:   18:        static int impl(U) { return 1; }
        -:   19:
        -:   20:        template <typename U = T, std::enable_if_t<!(std::is_same<U, char>::value && isSafe), int> = 0>
function int S<int>::impl<int, 0>(int) called 1 returned 100% blocks executed 100%
        1:   21:        static int impl(U) { return 1; }
        -:   22:
        -:   23:public:
        2:   24:        static int f(T t) {
        2:   25:                return impl(t);
        -:   26:        }
        -:   27:};
```
It would seem to look good here, but gcovr chokes on it.

When using the above return type SFINAE instead, then I get:
```
        -:    8:template <typename T>
        -:    9:class S
        -:   10:{
        -:   11:#if defined(USE_SAFE_FALLBACK)
        -:   12:        static constexpr bool isSafe = true;
        -:   13:#else
        -:   14:        static constexpr bool isSafe = false;
        -:   15:#endif
        -:   16:
        -:   17:        template <typename U = T>
        -:   18:        static
        -:   19:        std::enable_if_t<std::is_same<U, char>::value && isSafe, int>
        -:   20:        impl(U) { return 1; }
        -:   21:
        -:   22:        template <typename U = T>
        -:   23:        std::enable_if_t<!(std::is_same<U, char>::value && isSafe), int>
        -:   24:        static
        2:   25:        impl(U) { return 1; }
------------------
std::enable_if<!(std::is_same<int, char>::value&&S<int>::isSafe), int>::type S<int>::impl<int>(int):
function std::enable_if<!(std::is_same<int, char>::value&&S<int>::isSafe), int>::type S<int>::impl<int>(int) called 1 returned 100% blocks executed 100%
        1:   25:        impl(U) { return 1; }
------------------
std::enable_if<!(std::is_same<char, char>::value&&S<char>::isSafe), int>::type S<char>::impl<char>(char):
function std::enable_if<!(std::is_same<char, char>::value&&S<char>::isSafe), int>::type S<char>::impl<char>(char) called 1 returned 100% blocks executed 100%
        1:   25:        impl(U) { return 1; }
------------------
        -:   26:
        -:   27:public:
        2:   28:        static int f(T t) {
        2:   29:                return impl(t);
        -:   30:        }
        -:   31:};
```
where it looks as if gcov attributes the implementation of line 20 (T==char) to line 25 (T!=char), and then shows line 25 twice, but with different signature.
gcovr has no troubles generating the report for this one.

But that might be a red herring, as I have also seen, in my experiments, a similar output from the non-working code, namely:
```
        -:    8:template <typename T>
        -:    9:class S
        -:   10:{
        -:   11:#if defined(USE_SAFE_FALLBACK)
        -:   12:        static constexpr bool isSafe = true;
        -:   13:#else
        -:   14:        static constexpr bool isSafe = false;
        -:   15:#endif
        -:   16:
        -:   17:        // gcovr 5.2 crashes with this approach:
        -:   18:        template <typename U = T, std::enable_if_t<std::is_same<U, char>::value && isSafe, int> = 0>
        -:   19:        static int impl(U) { return 1; }
        -:   20:
        -:   21:        template <typename U = T, std::enable_if_t<!(std::is_same<U, char>::value && isSafe), int> = 0>
        2:   22:        static int impl(U) { return 1; }
------------------
int S<int>::impl<int, 0>(int):
function int S<int>::impl<int, 0>(int) called 1 returned 100% blocks executed 100%
        1:   22:        static int impl(U) { return 1; }
------------------
int S<char>::impl<char, 0>(char):
function int S<char>::impl<char, 0>(char) called 1 returned 100% blocks executed 100%
        1:   22:        static int impl(U) { return 1; }
------------------
        -:   23:
        ...
```
This one, again, lets gcovr abort with that "assert left.lineno == right.lineno" error.

So we have, essentially, very similar gcov output that works with the return type SFINAE but not with the template non-type argument SFINAE.

**Comment by Spacetown** [comment]
@mawillers I can reproduce it but I have no idea how to workaround this internal in gcovr. The functions are realy defined at different lines. If #663 is merged w can add a option to turn of this assertion. Until this you can do following workaround:
- Define both functions in the same line.
- Exclude the whole file with the option `--exclude`.

**Comment by fortysixandtwo** [comment]
Just wanted to leave another data point from a GObject based project:

```
#ifdef FOR_TESTING

G_DEFINE_TYPE_WITH_CODE
  (CallsDummyProvider, calls_dummy_provider, CALLS_TYPE_PROVIDER,
  G_IMPLEMENT_INTERFACE (CALLS_TYPE_MESSAGE_SOURCE,
                         calls_dummy_provider_message_source_interface_init))

#else

G_DEFINE_DYNAMIC_TYPE_EXTENDED
  (CallsDummyProvider, calls_dummy_provider, CALLS_TYPE_PROVIDER, 0,
   G_IMPLEMENT_INTERFACE_DYNAMIC (CALLS_TYPE_MESSAGE_SOURCE,
                                  calls_dummy_provider_message_source_interface_init))

#endif /* FOR_TESTING */
```

In one compilation unit I compile an executable for unit testing internal API and in the other case I end up with a module that is loaded during runtime.

I'm excluding these files for now.

Thanks for maintaining gcovr!

**Comment by fortysixandtwo** [comment]
@mawillers 

> Setting that `ignore_function_lineno` flag to `True` appears to fix it for me, too.

Did you mean you changed this in `merging.py` or is that something that can be set in `gcovr.cfg`?

**Comment by mawillers** [comment]
@fortysixandtwo

> > Setting that `ignore_function_lineno` flag to `True` appears to fix it for me, too.
> 
> Did you mean you changed this in `merging.py` or is that something that can be set in `gcovr.cfg`?

For that test, I patched `merging.py`, as there does not seem to be any setting (yet?) in `gcovr.cfg` for it.

As working with a modified `gcovr` is quite a hassle in my environment, my "real" fix so far is the aforementioned change of my own C++ implementation, though, so that it does not trigger this issue anymore.

**Comment by ubitux** [comment]
We are affected by this issue in the FFmpeg project as well, in these 3 different places at least:
- https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/aacsbr_template.c with `sbr_qmf_analysis`
- https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/fft_template.c with `fft_calc_c`
- https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/mpegaudiodsp_template.c with `round_sample`

As implied by the name, it's always because of "C templating" where the file is included with a particular configuration.

**Comment by ubitux** [comment]
> Can someone check the branch of #700 in his environment?

Seems to work with `--merge-mode-functions merge-use-line-0` with FFmpeg, thanks.

Thought I don't think it's a good idea to default to `strict`. From a user experience having random assertion errors enabled by default is pretty terrible, especially if that assert can be triggered on known user input.

**Comment by Spacetown** [comment]
> Thought I don't think it's a good idea to default to `strict`. From a user experience having random assertion errors enabled by default is pretty terrible, especially if that assert can be triggered on known user input.

Most users don't have the problem, therefore I want to keep the old behavior as default. If someone accepts it he has to decide how to do it. I prefer the value `separate`.

**Comment by ubitux** [comment]
> > Thought I don't think it's a good idea to default to `strict`. From a user experience having random assertion errors enabled by default is pretty terrible, especially if that assert can be triggered on known user input.
> 
> Most users don't have the problem, therefore I want to keep the old behavior as default. If someone accepts it he has to decide how to do it. I prefer the value `separate`.

Still, those rare users (I'm counting more than a handful in this issue alone, and not everyone is using gcovr in their project) are not at fault: the code is valid C. It's also a regression since it wasn't triggered before.

**Comment by Spacetown** [comment]
Two hands of thousands.
Merge them together was wrong and no one recognized it before. The new parser also discovered a GCC bug with negative counters, should this also be ignored?
This are really two different functions. @latk what do you think about the default?

**Comment by wangyoucao577** [comment]
> We are affected by this issue in the FFmpeg project as well, in these 3 different places at least:
> 
> * https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/aacsbr_template.c with `sbr_qmf_analysis`
> * https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/fft_template.c with `fft_calc_c`
> * https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/mpegaudiodsp_template.c with `round_sample`
> 
> As implied by the name, it's always because of "C templating" where the file is included with a particular configuration.

Met same issue on ffmpeg. Finally I [ignore these template files as workaround](https://github.com/wangyoucao577/ffmpeg-build/blob/main/.github/workflows/regression-tests.yml#L68):     
`gcovr -v --exclude '.*template.c' --html-details --output coverage/index.html`          

Even though the `include "xxx_template.c"` may not be good practicing from today's point of view, merge them together seems correct from semantic perspective. However, it obviously increases complexity and seems encouraged bad behavior. Maybe let it be is enough, and some `linter` program could tell people to change/ignore the old code properly.

**Comment by tsondergaard** [comment]
I've triggered `assert self.lineno == other.lineno` using gcovr 5.2 on AlmaLinux 9.1. By instrumenting gcovr I found out that it was the file below that it was struggling with where gcovr had line 17 and line 18 triggering the assertion. The problem went away when I simply moved the function definition at line 18 up to line 8.

```
        -:    0:Source:../evbase/include/evbase/UserDefinedStringLiterals.h
        -:    0:Graph:/home/ts/src/easyviz-2/build/sturgeon/src/CMakeFiles/sturgeon.dir/LoupeTools.cc.gcno
        -:    0:Data:/home/ts/src/easyviz-2/build/sturgeon/src/CMakeFiles/sturgeon.dir/LoupeTools.cc.gcda
        -:    0:Runs:30
        -:    1:// emacs: -*- Mode: c++; c-file-style: "stroustrup"; c-basic-offset: 4; indent-tabs-mode: nil; -*-
        -:    2:#ifndef EVBASE_USER_DEFINED_STRING_LITERALS_H
        -:    3:#define EVBASE_USER_DEFINED_STRING_LITERALS_H
        -:    4:
        -:    5:#include <QString>
        -:    6:#include <QChar>
        -:    7:
        -:    8:constexpr inline QLatin1Char operator "" _qlc(char c) noexcept
        -:    9:{
        -:   10:    return QLatin1Char(c);
        -:   11:}
        -:   12:
        -:   13:constexpr inline QLatin1String operator "" _qls(const char *s, std::size_t len) noexcept
        -:   14:{
        -:   15:    return QLatin1String(s, len);
        -:   16:}
function operator"" _qs(char const*, unsigned long) called 0 returned 0% blocks executed 0%
    #####:   17:
        -:   18:inline QString operator "" _qs(const char *s, std::size_t len) noexcept
    #####:   19:{
call    0 never executed
        -:   20:    return QString::fromUtf8(s, len);
        -:   21:}
        -:   22:
        -:   23:#endif // EVBASE_USER_DEFINED_STRING_LITERALS_H
```

I'm wondering about that blank line 17 in 

```
function operator"" _qs(char const*, unsigned long) called 0 returned 0% blocks executed 0%
    #####:   17:
        -:   18:inline QString operator "" _qs(const char *s, std::size_t len) noexcept
```

When I look in all the other gcov files, it doesn't look like that.

Information about my system:

```
$ cat /etc/redhat-release 
AlmaLinux release 9.1 (Lime Lynx)

$ gcovr --version
gcovr 5.2

[...]

$ gcc --version
gcc (GCC) 11.3.1 20220421 (Red Hat 11.3.1-2)

[...]

$ ld.gold --version
GNU gold (version 2.35.2-24.el9) 1.16

[...]

$ gcov --version
gcov (GCC) 11.3.1 20220421 (Red Hat 11.3.1-2)

[...]
```

**Comment by Spacetown** [comment]
@tsondergaard Sorry, we can't help you with the problem why a blank line is reported by gcov in some files and not in others. This is related to gcov and not gcovr.

**Comment by tsondergaard** [comment]
@Spacetown, yep, I'm merely reporting the observation here in the hope that it can help the gcovr developers make gcovr robust with regards to the output produced by gcov. My comment about the blank line was just an observation that I again just thought might be relevant.

I also hope the comment may be helpful to other gcovr users - since I think it is quite unexpected that the assert, in some cases at last, can be avoided simply by reordering functions in a source file.

**Comment by Spacetown** [comment]
On current master branch a function merge mode is available. I'm waiting for a last PR to be closed before creating the release 6.0.
